Entity listener utilities for ktx-ashley; typos

RedGrapefruit09 opened the pull request:
Resolves #409 .

- [x] `EntityAdditionListener`
- [x] `EntityRemovalListener`
- [x] Lambda utilities
- [x] KDocs
- [x] Tests
- [x] Changelog and `README.md` updates

czyzby commented on ashley/src/main/kotlin/ktx/ashley/entity_listeners.kt:
```diff
@@ -0,0 +1,26 @@
+package ktx.ashley
```
Not sure if underscores are the standard in Kotlin file names. You can rename the file to just `listeners.kt`.

RedGrapefruit09 commented:
@czyzby , is this the proper way of implementing your suggestion with lambdas? (below)

```kotlin
inline fun EntitySystem.onEntityAdded(action: (entity: Entity) -> Unit) {
  engine.addEntityListener(/* some impl of EntityAdditionListener that takes a lambda? */ null)
}
```

czyzby commented on ashley/src/main/kotlin/ktx/ashley/listeners.kt:
```diff
@@ -0,0 +1,26 @@
+/**
+ * An [EntityListener] only listening to the [entityAdded] event.
+ *
+ * A [Unit] implementation is provided for the [entityRemoved] event, as it is ignored.
```
`[Unit] implementation` might not be clear for the users. I'd go with a `no-op` or `empty implementation`.

RedGrapefruit09 replied:
Fixed in https://github.com/libktx/ktx/pull/410/commits/a2cdbcdf9ebcd52bdb467d5ad43b43e94096e964!
I chose to go with `empty implementation` as it seems less ambiguous to me.

RedGrapefruit09 commented:
> If that's OK, please also update the changelog.

Done! I also updated the docs and added examples in the `README.md` of the `ktx-ashley` module.
I'll do this again once the lambda-related changes are finished. Thanks for your feedback, once again!

RedGrapefruit09 commented on ashley/src/main/kotlin/ktx/ashley/entity_listeners.kt:
```diff
@@ -0,0 +1,26 @@
+ *
+ * If you need both events from the [EntityListener], implement the regular [EntityListener] from Ashley.
+ */
+interface EntityAdditionListener : EntityListener {
```
Done in https://github.com/libktx/ktx/pull/410/commits/9d676a84ed2bc75b96aa211bcde09108bd22cb6b!

RedGrapefruit09 commented:
I accidentally put the changelog changes in the `1.10.0-rc1` header, not in the `1.10.0-SNAPSHOT` header. I fixed it in https://github.com/libktx/ktx/pull/410/commits/951e9e777c91065d95e7004b26e30d0305c491a6.

RedGrapefruit09 commented on ashley/src/test/kotlin/ktx/ashley/listenerTest.kt:
```diff
@@ -0,0 +1,17 @@
+class `should implement EntityAdditionListener with only one method implemented` : EntityAdditionListener {
+  // Guarantees EntityAdditionListener's entityRemoved method is always optional to implement
+
+  override fun entityAdded(entity: Entity?) = Unit
```
I'm not sure if the `entity` is always not null, for now, I'll change the definition in the interfaces to be `Entity?` in the next commit.

RedGrapefruit09 replied:
Resolved in new commits.

RedGrapefruit09 commented:
> Change the method signature to return an `EntityListener` (or even an `EntityAdditionListener`/`EntityRemovalListener`). Return the created instance of the listener, allowing the user to save a reference to it and eventually remove it from the engine.

Ok, I'll do that.

RedGrapefruit09 commented:
I made a prototype of the lambda utils (see https://github.com/libktx/ktx/pull/410/commits/cd8331ee7060525a770b85fd46a05527983415fc), but I, for some reason, I can't do this:

```kotlin
  // this compiles
  engine.onEntityAdded({
    
  })
  
  // this doesn't compile
  engine.onEntityAdded {
    
  }
```

czyzby commented:
You have to put the lambda as the last parameter, after the family and priority.

RedGrapefruit09 commented:
@czyzby I've finished implementing the lambda utils and added KDocs.
Are there any changes left to be made until I've got to update the changelog and README?

czyzby commented:
As far as I can tell, there are only some issues related to the naming of variables and such. `emptyFamily` sounds more like `anyFamily` or `allFamilies`, since it matches every family as far as I can tell. I'd also rename the `implementation` lambda parameters to something like `listener` so that they are clearer if someone decides to use the named parameters.

RedGrapefruit09 commented:
@czyzby I've implemented your feedback in https://github.com/libktx/ktx/pull/410/commits/0625d710a45121ba6e52989e8130d7382a70f361.
I'll make some tests (`ktx-ashley` seems to use this Spek thing, so I guess I'll learn it) later.

RedGrapefruit09 commented:
I've also now updated the changelog and added some docs.

czyzby commented:
About the tests: `ktx-ashley` was contributed as a whole with Spek tests and I didn't see an issue with that at the time, even though most modules use JUnit. Nowadays, I'd prefer if all tests used the same technology (and structure) throughout the project, even if that means going with a less modern approach.

If you have trouble writing Spek tests, I'll accept the PR with JUnit test suite as well. However, Spek would be more consistent with the current state of this module. I might consider rewriting the tests to JUnit some time, even though I personally don't have any problems with the Spek framework - it's just easier to rewrite 1 module for consistency rather than everything else.

RedGrapefruit09 commented:
I've made Spek tests that (I think) cover everything, so I think this PR is ready for review.

czyzby commented on ashley/src/main/kotlin/ktx/ashley/listeners.kt:
```diff
@@ -0,0 +1,124 @@
+ * If you need both events from the [EntityListener], implement the regular [EntityListener] from Ashley.
+ */
+interface EntityAdditionListener : EntityListener {
+  override fun entityAdded(entity: Entity)
```
The only thing that comes to mind at this point is that we might want to include some documentation for the methods that should be overridden, i.e. something along the lines of `Invoked when an matching [Entity] is added to the engine. See [EntityListener.entityAdded].`.

RedGrapefruit09 replied:
Resolved in https://github.com/libktx/ktx/pull/410/commits/56811de1b8d027612a7422f7a36cc53205a68cf5.

czyzby reviewed (approved):
Looks great, thanks for your contribution. I'll go through this again and merge it later today.

czyzby commented:
Thanks for the contribution, @RedGrapefruit09!
